### PR TITLE
Defer mr_cache_init until first mr_reg, delay closing MR after rendezvous send

### PR DESCRIPTION
**Pull request by ztiffany**
Delaying mr_cache_init until after a domain is created allows us to tune MR
cache parameters with domain provider ops.

Signed-off-by: Zach Tiffany ztiffany@cray.com

@jswaro @hppritcha @sungeunchoi I am going towards creating a test which uses lazy deregistration.  The first step is allowing us to configure a domain's mr cache attributes.

### CODE REVIEW

**Comment by ztiffany** [comment]
The second commit here adds the GNI_MR_CACHE_LAZY_DEREG domain op parameter.  It's used in a new rdm_sr test which reproduces issue #469.

The second commit fixes issue #469 by delaying closing the MR used for sends until outside the SMSG RX callback.

**Comment by ztiffany** [comment]
Updated commit with fixed error handling in mr_cache_init().

File: prov/gni/src/gnix_mr.c [hunk 1]
```diff
@@ -554,44 +566,54 @@ int _gnix_mr_cache_init(
 		cache_attr = attr;
 	}
 
+	cache_p = (gnix_mr_cache_t *)calloc(1, sizeof(*cache_p));
```
Comment (jswaro) [inline prov/gni/src/gnix_mr.c]: Space between the cast and the calloc please. Just a nit

**Comment by jswaro** [comment]
:+1:

File: prov/gni/src/gnix_mr.c [hunk 2]
```diff
@@ -386,6 +386,10 @@ uint64_t _gnix_convert_mhdl_to_key(gni_mem_handle_t *mhdl)
 	return key.value;
 }
 
+static int __gnix_mr_cache_init(
+		gnix_mr_cache_t **cache,
+		gnix_mr_cache_attr_t *attr);
+
```
Comment (hppritcha) [inline prov/gni/src/gnix_mr.c]: would you mind putting this prototype up near the top of the file?

**Comment by hppritcha** [comment]
assuming minor things are taken care of, :+1:

**Pull request merged**